File: OOPADS-dissection/FunctionalMath.py
from functools import reduce
import math
import random

def main():
    # Lambda Expressions
    add = lambda x, y: x + y
    print("Addition:", add(5, 3))

    # Functional Programming with List Comprehensions and Filters
    numbers = [1, 2, 3, 4, 5, 6]
    filtered = list(filter(lambda x: x % 2 == 0, numbers))
    print("Filtered Even Numbers:", filtered)

    # Using map with Math functions
    squares = list(map(math.sqrt, numbers))
    print("Square Roots:", squares)

    # Higher-Order Functions
    print("Custom Operation (Multiplication):",
          apply_operation(6, 7, lambda a, b: a * b))

    # Recursion
    print("Factorial of 5:", factorial(5))

    # Lazy Evaluation with Generators
    random_numbers = (random.random() for _ in range(5))
    print("Random Numbers:")
    for num in random_numbers:
        print(num)

def apply_operation(x, y, operation):
    return operation(x, y)

def factorial(n):
    if n in (0, 1):
        return 1
    return n * factorial(n - 1)

if __name__ == "__main__":
    main()

File: OOPADS-dissection/FunctionalMath.java
import java.util.*;
import java.util.function.*;
import java.util.stream.*;

public class FunctionalMath {

    public static void main(String[] args) {
        // Example of Lambda Expression
        BiFunction<Integer, Integer, Integer> add = (x, y) -> x + y;
        System.out.println("Addition: " + add.apply(5, 3));

        // Functional Interface with Streams
        List<Integer> numbers = Arrays.asList(1, 2, 3, 4, 5, 6);
        List<Integer> filtered = numbers.stream()
                                        .filter(x -> x % 2 == 0)
                                        .collect(Collectors.toList());
        System.out.println("Filtered Even Numbers: " + filtered);

        // Method Reference
        List<Double> squares = numbers.stream()
                                       .map(Math::sqrt)
                                       .collect(Collectors.toList());
        System.out.println("Square Roots: " + squares);

        // Higher-Order Function
        System.out.println("Custom Operation (Multiplication): " +
                applyOperation(6, 7, (a, b) -> a * b));

        // Recursion
        System.out.println("Factorial of 5: " + factorial(5));

        // Lazy Evaluation
        Stream<Double> randomNumbers = Stream.generate(Math::random).limit(5);
        randomNumbers.forEach(System.out::println);
    }

    public static int applyOperation(int x, int y, BiFunction<Integer, Integer, Integer> operation) {
        return operation.apply(x, y);
    }

    public static int factorial(int n) {
        if (n == 0 || n == 1) {
            return 1;
        }
        return n * factorial(n - 1);
    }
}
